go/pkg/contextmd/contextmd.go

```diff
@@ -177,33 +177,33 @@ func mergeSet(set map[string]struct{}) string {
 
 // capToLimit ensures total length does not exceed max header size.
 func capToLimit(m *Metadata, limit int) *Metadata {
-	total := len(m.ToolName) + len(m.ToolVersion) + len(m.ActionID) + len(m.InvocationID) + len(m.CorrelatedInvocationID)
+	total := len(m.ToolName) + len(m.ToolVersion) + len(m.ActionID) + len(m.InvocationID) + len(m.CorrelatedInvocationsID)
 	excess := total - limit
 	if excess <= 0 {
 		return m
 	}
-	// We ignore the tool name, because in practice this is a
-	// very short constant which makes no sense to truncate.
-	diff := len(m.ActionID) - len(m.InvocationID)
-	if diff > 0 {
-		if diff > excess {
-			m.ActionID = m.ActionID[:len(m.ActionID)-excess]
-		} else {
-			m.ActionID = m.ActionID[:len(m.ActionID)-diff]
-			rem := (excess - diff + 1) / 2
-			m.ActionID = m.ActionID[:len(m.ActionID)-rem]
-			m.InvocationID = m.InvocationID[:len(m.InvocationID)-rem]
-		}
-	} else {
-		diff = -diff
-		if diff > excess {
-			m.InvocationID = m.InvocationID[:len(m.InvocationID)-excess]
-		} else {
-			m.InvocationID = m.InvocationID[:len(m.InvocationID)-diff]
-			rem := (excess - diff + 1) / 2
-			m.InvocationID = m.InvocationID[:len(m.InvocationID)-rem]
-			m.ActionID = m.ActionID[:len(m.ActionID)-rem]
+	truncateStrings(limit, &m.ActionID, &m.InvocationID, &m.CorrelatedInvocationsID, &m.ToolName, &m.ToolVersion)
+	return m
+}
+
+// truncateStrings remove one char from the longest str at a time until the
+// total length of all the strings is less than limit value.
+func truncateStrings(limit int, inputs ...*string) {
```

Comment:
Curious why you're passing strings by reference here? Strings in Go are backed by slices, which are cheap to copy because the underlying array is shared.

Reply:
I want to pass m's fields here, and modify them in place (basically modify m in place).

say, if I pass strings here, and the interface is `func truncateStrings(limit int, inputs ...string) outputs []string{}`

I will end up with very long code on the caller's side. I was keep thinking it over and over for about an hour this morning, didn't find out a better solution than pass in ...*string.

```
outputs := truncateStrings(32, m.a, m.b, m.c, m.d, m.e)
m.a = outputs[0]
m.b = outputs[1]
m.c = outputs[2]
m.d = outputs[3]
m.e = outputs[4]
return m
```